runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/graph/Networks.java

```diff
@@ -193,29 +201,55 @@ public final NodeT apply(NodeT input) {
         "Only networks without self loops are supported, given %s",
         network);
 
-    // Linked hashset will prevent duplicates from appearing and will maintain insertion order.
-    LinkedHashSet<NodeT> nodes = new LinkedHashSet<>(network.nodes().size());
-    Queue<NodeT> processingOrder = new ArrayDeque<>();
-    // Add all the roots
-    for (NodeT node : network.nodes()) {
-      if (network.inDegree(node) == 0) {
-        processingOrder.add(node);
-      }
-    }
+    // Uses the following algorithm:
```

Comment:
Only nodes that are part of cycles are ever removed as part of the "delta" calculation.

The algorithm:
1) Removes all sinks from the graph (including newly created sinks once the graph is updated) until there are none.
2) Removes all sources from the graph (including newly created sources once the graph is updated) until there are none.
3) Remote a single back edge which highest delta. This may break a cycle. 

Concretely, in your example, the algorithm would:
* add Sink to the beginning of S2, removing it from the graph.
* add B to the beginning of S2, removing it from the graph.
* add A to the beginning of S2, removing it from the graph.
* add Source1 or Source2 (ambiguous so lets say Source1) to the beginning of S2, removing it from the graph.
* add Source2 to the beginning of S2, removing it from the graph.


I expanded the comment within the code about the algorithm to clarify some parts of it.

Reply:
Ah, the detail I missed was that it the algorithm would gather any newly created sinks/sources (I thought after removing Sink node B wouldn't be considered a sink). It makes sense now and I believe it does work for topological ordering.